### PR TITLE
KEP-2400: Alpha node support for swap memory

### PR DESCRIPTION
**Pull request by ehashman**
KEP: https://github.com/kubernetes/enhancements/issues/2400

Implementation PR: https://github.com/kubernetes/kubernetes/pull/102823

### CODE REVIEW

**Comment by sftim** [comment]
A question about the feature gate name. Is it `NodeMemorySwap` or `NodeSwapEnabled` or both?

We usually don't put “Enable” or “Enabled” in feature gate names, so when I saw `NodeSwapEnabled` I was surprised. Typically you can take a feature gate name and use it to fill in the phrase “If you set this particular feature flag to true, it means that you enable _thing_ in your cluster”.

**Comment by ehashman** [comment]
https://github.com/kubernetes/kubernetes/commit/0dd4ce40adc8e2f51de6ca16b1f18a0e89503ee4

It's `NodeSwapEnabled`. That missed all of KEP review, PRR, and PR review. It'll be non-trivial to change the name at this point.

**Comment by sftim** [comment]
A nice-to-have PR to replace the feature gate name: https://github.com/kubernetes/kubernetes/pull/103553

**Comment by ehashman** [comment]
lol bot no!

/reopen

**Comment by llhuii** [comment]
> [kubernetes/kubernetes@0dd4ce4](https://github.com/kubernetes/kubernetes/commit/0dd4ce40adc8e2f51de6ca16b1f18a0e89503ee4)
> 
> It's `NodeSwapEnabled`. That missed all of KEP review, PRR, and PR review. It'll be non-trivial to change the name at this point.

It's still `NodeSwapEnabled` in [KEP-2400](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md#design-details), would you mind updating it?

**Comment by ehashman** [comment]
@llhuii there's an open PR but it hasn't been approved: https://github.com/kubernetes/enhancements/pull/2815

**Comment by llhuii** [comment]
@ehashman got it, sorry to miss that.

**Comment by PI-Victor** [comment]
Hello @ehashman  :wave: please take a look at [Documenting for a release](https://kubernetes.io/docs/contribute/new-content/new-features/#pr-ready-for-review) - PR Ready for Review to get your PR ready for review before Tuesday 20th of July, 11:59 PM PDT. Thank you!

File: content/en/docs/concepts/architecture/nodes.md [hunk 1]
```diff
@@ -392,6 +392,50 @@ Failed pod objects will be preserved until explicitly deleted or [cleaned up by
 This is a change of behavior compared to abrupt node termination.
 {{< /note >}}
 
+## Swap memory management {#swap-memory}
+
+{{< feature-state state="alpha" for_k8s_version="v1.22" >}}
+
+Prior to Kubernetes 1.22, nodes did not support the use of swap memory, and a
+kubelet would fail to start if swap was detected on a node by default. In 1.22
+onwards, swap memory support can be configured on a per-node basis.
```
Comment (tengqm) [inline content/en/docs/concepts/architecture/nodes.md:401]:
```suggestion
kubelet would by default fail to start if swap was detected on a node. In 1.22
onwards, swap memory support can be enabled on a per-node basis.
```

File: content/en/docs/concepts/architecture/nodes.md [hunk 2]
```diff
@@ -392,6 +392,50 @@ Failed pod objects will be preserved until explicitly deleted or [cleaned up by
...
+
+To enable swap on a node, the `NodeSwap` feature flag must be set to true on
```
Comment (tengqm) [inline content/en/docs/concepts/architecture/nodes.md:403]:
```suggestion
To enable swap on a node, the `NodeSwap` feature gate must be enabled on
```

File: content/en/docs/concepts/architecture/nodes.md [hunk 3]
```diff
@@ -392,6 +392,50 @@ Failed pod objects will be preserved until explicitly deleted or [cleaned up by
...
+the kubelet, and the `--fail-swap-on` command line flag or `failSwapOn`
+[configuration setting] must be set to false.
+
+[configuration setting]: https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/#kubelet-config-k8s-io-v1beta1-KubeletConfiguration
```
Comment (tengqm) [inline content/en/docs/concepts/architecture/nodes.md:407]:
```suggestion
[configuration setting](/docs/reference/config-api/kubelet-config.v1beta1/#kubelet-config-k8s-io-v1beta1-KubeletConfiguration)
must be set to false.
```

File: content/en/docs/concepts/architecture/nodes.md [hunk 4]
```diff
@@ -392,6 +392,50 @@ Failed pod objects will be preserved until explicitly deleted or [cleaned up by
...
+
+A user can also optionally configure `memorySwap.swapBehavior` in order to
+configure how the node will use swap memory. For example,
```
Comment (tengqm) [inline content/en/docs/concepts/architecture/nodes.md:410]:
```suggestion
specify how a node will use swap memory. For example,
```

Just avoid duplicated "configure" verbs.

File: content/en/docs/concepts/architecture/nodes.md [hunk 5]
```diff
@@ -392,6 +392,50 @@ Failed pod objects will be preserved until explicitly deleted or [cleaned up by
...
+
+```yaml
+memorySwap:
+  swapBehavior: LimitedSwap
+```
+
+The available configuration options for `SwapBehavior` are:
+
+- Unset: defaults to `LimitedSwap`
```
Comment (tengqm) [inline content/en/docs/concepts/architecture/nodes.md:419]: Maybe we want to remove this line or clarify what "Unset" means. Does it mean that the "swapBehavior" key is omitted or the "swapBehavior" is set to `null`, or both?

File: content/en/docs/concepts/architecture/nodes.md [hunk 6]
```diff
@@ -392,6 +392,50 @@ Failed pod objects will be preserved until explicitly deleted or [cleaned up by
...
+
+A user can also optionally configure `memorySwap.swapBehavior` in order to
+configure how the node will use swap memory. For example,
+
+```yaml
+memorySwap:
+  swapBehavior: LimitedSwap
+```
+
+The available configuration options for `SwapBehavior` are:
```
Comment (tengqm) [inline content/en/docs/concepts/architecture/nodes.md:417]:
```suggestion
The available configuration options for `swapBehavior` are:
```

File: content/en/docs/concepts/architecture/nodes.md [hunk 7]
```diff
@@ -392,6 +392,50 @@ Failed pod objects will be preserved until explicitly deleted or [cleaned up by
...
+- `UnlimitedSwap`: Kubernetes workloads can use as much swap memory as they
+  request, up to the system limit.
+
+The behaviour of the `LimitedSwap` setting depends if the node is running with
+v1 or v2 of control groups (also known as "cgroups"):
+
+|           | `LimitedSwap`                                                                                          | `UnlimitedSwap`                                                             |
+|-----------|--------------------------------------------------------------------------------------------------------|-----------------------------------------------------------------------------|
+| cgroupsv1 | Kubernetes workloads can use any combination of memory and swap, up to the pod's memory limit, if set. | Kubernetes workloads can use unlimited swap memory, up to the system limit. |
+| cgroupsv2 | Kubernetes workloads cannot use swap memory.                                                           | Kubernetes workloads can use unlimited swap memory, up to the system limit. |
```
Comment (tengqm) [inline content/en/docs/concepts/architecture/nodes.md:431]: Please consider revise the table into a simple list:

- Markdown tables are very limited in terms of table formatting, e.g. you cannot break lines freely, you cannot embed lists in cells etc.
- The `UnlimitedSwap` behavior has been clarified above, so no need to repeat it here.

File: content/en/docs/concepts/architecture/nodes.md [hunk 8]
```diff
@@ -392,6 +392,50 @@ Failed pod objects will be preserved until explicitly deleted or [cleaned up by
...
+
+For more information, and to assist with testing and provide feedback, please
+see [KEP-2400] and its [design proposal].
+
+[KEP-2400]: https://github.com/kubernetes/enhancements/issues/2400
+[design proposal]: https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md
```
Comment (tengqm) [inline content/en/docs/concepts/architecture/nodes.md:437]: Please try use inline links when possible. We have link error detection script that cannot handle the out-of-band flavor.

```suggestion
see [KEP-2400](https://github.com/kubernetes/enhancements/issues/2400) and its
[design proposal](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/2400-node-swap/README.md).
```

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 9]
```diff
@@ -783,6 +784,8 @@ Each feature gate is designed for enabling/disabling a specific feature:
 - `NodeDisruptionExclusion`: Enable use of the Node label `node.kubernetes.io/exclude-disruption`
   which prevents nodes from being evacuated during zone failures.
 - `NodeLease`: Enable the new Lease API to report node heartbeats, which could be used as a node health signal.
+- `NodeSwap`: Enable the kubelet to allocate swap memory for Kubernetes workloads on a node. Must be used with `KubeletConfiguration.FailSwapOn` set to false.
```
Comment (tengqm) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:787]: Not sure if the spelling is correct, but wrapping long lines is always a good habit.

```suggestion
- `NodeSwap`: Enable the kubelet to allocate swap memory for Kubernetes workloads on a node.
  Must be used with `KubeletConfiguration.failSwapOn` set to false.
```

File: content/en/docs/concepts/architecture/nodes.md (hunk 5 above)
Comment (ehashman) [reply content/en/docs/concepts/architecture/nodes.md:419]: Both.

Comment (ehashman) [reply content/en/docs/concepts/architecture/nodes.md]: Moved this to a trailing paragraph.

**Comment by sftim** [comment]
LGTM

I think we want this to have tech review from SIG Node / someone who's confident they can provide accurate feedback.

**Comment by ehashman** [comment]
/assign @mrunalp 
SIG Node technical reviewer

**Comment by tengqm** [comment]
This one is good to go.
/approve

**Pull request merged**